sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java

```diff
@@ -797,6 +854,61 @@ public Mono<List<Address>> getMasterAddressesViaGatewayAsync(
         });
     }
 
+    /***
+     *  merge the new addresses get back from gateway with the cached addresses.
+     *  If the address is being returned from gateway again, then keep using the cached addressInformation object
+     *  If it is a new address being returned, then use the new addressInformation object.
+     *
+     * @param newAddresses the latest addresses being returned from gateway.
+     * @param cachedAddresses the cached addresses.
+     *
+     * @return the merged addresses.
+     */
+    private AddressInformation[] mergeAddresses(AddressInformation[] newAddresses, AddressInformation[] cachedAddresses) {
+        checkNotNull(newAddresses, "Argument 'newAddresses' should not be null");
+
+        if (cachedAddresses == null) {
+            return newAddresses;
+        }
+
+        List<AddressInformation> mergedAddresses = new ArrayList<>();
+        Map<Uri, AddressInformation> cachedAddressMap =
+                Arrays
+                    .stream(cachedAddresses)
+                    .collect(Collectors.toMap(address -> address.getPhysicalUri(), address -> address));
+
+        for (AddressInformation addressInformation : newAddresses) {
+            if (cachedAddressMap.containsKey(addressInformation.getPhysicalUri())) {
+                mergedAddresses.add(cachedAddressMap.get(addressInformation.getPhysicalUri()));
+            } else {
+                mergedAddresses.add(addressInformation);
+            }
+        }
+
+        return mergedAddresses.toArray(new AddressInformation[mergedAddresses.size()]);
+    }
+
+    private void validateReplicaAddresses(AddressInformation[] addresses) {
+        checkNotNull(addresses, "Argument 'addresses' can not be null");
+
+        // By theory, when we reach here, the status of the address should be in one of the three status: Unknown, Connected, UnhealthyPending
+        // using open connection to validate addresses in UnhealthyPending status
+        // Could extend to also open connection for unknown in the future
+        List<Uri> addressesNeedToValidation =
+                Arrays
+                    .stream(addresses)
+                    .map(address -> address.getPhysicalUri())
+                    .filter(addressUri -> addressUri.getHealthStatus() == Uri.HealthStatus.UnhealthyPending)
+                    .collect(Collectors.toList());
+
+        if (addressesNeedToValidation.size() > 0) {
+            this.openConnectionsHandler
+                    .openConnections(addressesNeedToValidation)
```

Comment:
@xinlian12 I looked at the openConnections code in `RntbdOpenConnectionsHandler` class, and I think it can cause thread starvation. Let's discuss it offline since that class is not touched as part of this PR.